File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.js
/**
 * Finds the minimum number of moves to make all pairs sum to the same value
 * @param {number[]} nums - Array of integers
 * @param {number} limit - Maximum value any element can be changed to (1 to limit)
 * @returns {number} Minimum number of moves required
 */
function minMoves(nums, limit) {
    const arrayLength = nums.length;

    // Difference array to track cost changes at different sum values
    // Size is 2*limit+2 to handle all possible sums (2 to 2*limit)
    const differenceArray = Array(limit * 2 + 2).fill(0);

    // Process each pair from both ends of the array
    for (let i = 0; i < (arrayLength >> 1); ++i) {
        // Get the pair elements from opposite ends
        const minValue = Math.min(nums[i], nums[arrayLength - 1 - i]);
        const maxValue = Math.max(nums[i], nums[arrayLength - 1 - i]);

        // [2, minValue]: requires 2 moves
        differenceArray[2] += 2;
        differenceArray[minValue + 1] -= 2;

        // [minValue+1, minValue+maxValue-1]: requires 1 move
        differenceArray[minValue + 1] += 1;
        differenceArray[minValue + maxValue] -= 1;

        // [minValue+maxValue+1, maxValue+limit]: requires 1 move
        differenceArray[minValue + maxValue + 1] += 1;
        differenceArray[maxValue + limit + 1] -= 1;

        // [maxValue+limit+1, 2*limit]: requires 2 moves
        differenceArray[maxValue + limit + 1] += 2;
    }

    // Find the minimum cost using prefix sum
    let answer = arrayLength;
    let currentCost = 0;

    for (let targetSum = 2; targetSum < differenceArray.length; ++targetSum) {
        currentCost += differenceArray[targetSum];
        answer = Math.min(answer, currentCost);
    }

    return answer;
}

File: 1674-minimum-moves-to-make-array-complementary/1674-minimum-moves-to-make-array-complementary.java
class Solution {
    public int minMoves(int[] nums, int limit) {
        int n = nums.length;
        // 최대 합은 2 * limit이므로 2 * limit + 2 크기의 차분 배열 생성
        int[] diff = new int[2 * limit + 2];
        
        for (int i = 0; i < n / 2; i++) {
            int a = nums[i];
            int b = nums[n - 1 - i];
            
            // 1. 현재 쌍으로 만들 수 있는 최소/최대 합 (0번 이동)
            int minSum = Math.min(a, b) + 1;
            int maxSum = Math.max(a, b) + limit;
            int currentSum = a + b;
            
            // 기본: 모든 쌍은 2번 이동으로 목표 합을 만들 수 있음
            diff[2] += 2;
            diff[2 * limit + 1] -= 2;
            
            // 2. 1번 이동으로 가능한 범위: [minSum, maxSum]
            // 이 범위 내에서는 1번 이동(diff[minSum]~diff[maxSum])으로 변경 가능
            diff[minSum] -= 1;
            diff[maxSum + 1] += 1;
            
            // 3. 0번 이동(변경 불필요) 가능한 합: currentSum
            // 위에서 1번 이동 범위에 포함된 currentSum을 0번 이동으로 보정
            diff[currentSum] -= 1;
            diff[currentSum + 1] += 1;
        }
        
        int minMoves = n;
        int currentMoves = 0;
        // 차분 배열을 누적합으로 계산하여 최솟값 찾기
        for (int i = 2; i <= 2 * limit; i++) {
            currentMoves += diff[i];
            minMoves = Math.min(minMoves, currentMoves);
        }
        
        return minMoves;
    }
}
